windows/lcow_runtime.go

```diff
@@ -0,0 +1,400 @@
+	eventsapi "github.com/containerd/containerd/api/services/events/v1"
+	containerdtypes "github.com/containerd/containerd/api/types"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/events"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/runtime"
+	"github.com/containerd/containerd/typeurl"
+	"github.com/containerd/containerd/windows/hcsshimopts"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+const (
+	lcowRuntimeName = "lcow"
+)
+
+var (
+	lcowPluginID = fmt.Sprintf("%s.%s", plugin.RuntimePlugin, lcowRuntimeName)
+)
+
+var _ = (runtime.Runtime)(&LCOWRuntime{})
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		ID:   lcowRuntimeName,
+		Type: plugin.RuntimePlugin,
+		Init: NewLCOWPlugin,
+		Requires: []plugin.PluginType{
+			plugin.MetadataPlugin,
+		},
+	})
+}
+
+func NewLCOWPlugin(ic *plugin.InitContext) (interface{}, error) {
+	if err := os.MkdirAll(ic.Root, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create state directory at %s", ic.Root)
+	}
+
+	m, err := ic.Get(plugin.MetadataPlugin)
+	if err != nil {
+		return nil, err
+	}
+
+	cd := filepath.Join(ic.Root, cacheDirectory)
+	sd := filepath.Join(ic.Root, scratchDirectory)
+
+	// Make sure the scratch directory is created under dataRoot
+	if err := os.MkdirAll(sd, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create scratch directory at %s", ic.Root)
+	}
+
+	// Make sure the cache directory is created under dataRoot
+	if err := os.MkdirAll(cd, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create cache directory at %s", ic.Root)
+	}
+
+	r := &LCOWRuntime{
+		root:    ic.Root,
+		pidPool: newPidPool(),
+
+		events:    make(chan interface{}, 4096),
+		publisher: ic.Events,
+		// TODO(mlaventure): lcow needs a stat monitor
+		monitor: nil,
+		tasks:   runtime.NewTaskList(),
+		db:      m.(*bolt.DB),
+
+		cachedSandboxFile: filepath.Join(cd, sandboxFilename),
+		cachedScratchFile: filepath.Join(cd, scratchFilename),
+		cache:             make(map[string]*cacheItem),
+		serviceVms:        make(map[string]*serviceVMItem),
+		globalMode:        false,
+	}
+
+	// Load our existing containers and kill/delete them. We don't support
+	// reattaching to them
+	r.cleanup(ic.Context)
+
+	return r, nil
+}
+
+type LCOWRuntime struct {
+	sync.Mutex
+
+	root    string
+	pidPool *pidPool
+
+	publisher events.Publisher
+	events    chan interface{}
+
+	monitor runtime.TaskMonitor
+	tasks   *runtime.TaskList
+	db      *bolt.DB
+
+	cachedSandboxFile  string                    // Location of the local default-sized cached sandbox.
+	cachedSandboxMutex sync.Mutex                // Protects race conditions from multiple threads creating the cached sandbox.
+	cachedScratchFile  string                    // Location of the local cached empty scratch space.
```

Comment:
Ditto above about name.

The scratch is a VHD used by Utility VMs as disk space to perform operations that need extra a tmp dir for intermediate files such as producing diffs of the current layer. It is managed by the clients so that it can be created once and reused for each Utility VM.

Reply:
Thanks for the explanation, may be worth adding it to the comment

Reply:
Added.